crates/pop-common/src/sourcing/binary.rs

```diff
@@ -0,0 +1,253 @@
+	///
+	/// # Arguments
+	/// * `release` - Whether any binaries needing to be built should be done so using the release profile.
+	/// * `status` - Used to observe status updates.
+	/// * `verbose` - Whether verbose output is required.
+	pub async fn source(
+		&self,
+		release: bool,
+		status: &impl Status,
+		verbose: bool,
+	) -> Result<(), Error> {
+		match self {
+			Self::Local { name, path, manifest, .. } => match manifest {
+				None => {
+					return Err(Error::MissingBinary(format!(
+						"The {path:?} binary cannot be sourced automatically."
+					)))
+				},
+				Some(manifest) => {
+					from_local_package(manifest, name, release, status, verbose).await
+				},
+			},
+			Self::Source { source, cache, .. } => {
+				source.source(cache, release, status, verbose).await
+			},
+		}
+	}
+
+	/// Whether any locally cached version can be replaced with a newer version.
+	pub fn stale(&self) -> bool {
+		// Only binaries sourced from GitHub release archives can currently be determined as stale
+		let Self::Source { source: GitHub(ReleaseArchive { tag, latest, .. }), .. } = self else {
+			return false;
+		};
+		latest.as_ref().map_or(false, |l| tag.as_ref() != Some(l))
+	}
+
+	/// Specifies that the latest available versions are to be used (where possible).
+	pub fn use_latest(&mut self) {
+		if let Self::Source { source: GitHub(ReleaseArchive { tag, latest, .. }), .. } = self {
+			if let Some(latest) = latest {
+				*tag = Some(latest.clone())
+			}
+		};
+	}
+
+	/// If applicable, the version of the binary.
+	pub fn version(&self) -> Option<&str> {
+		match self {
+			Self::Local { .. } => None,
+			Self::Source { source, .. } => match source {
+				Git { reference, .. } => reference.as_ref(),
+				GitHub(source) => match source {
+					ReleaseArchive { tag, .. } => tag.as_ref(),
+					SourceCodeArchive { reference, .. } => reference.as_ref(),
+				},
+				Archive { .. } | Source::Url { .. } => None,
+			},
+		}
+		.map(|r| r.as_str())
+	}
+}
+
+/// A descriptor of a remote repository.
+#[derive(Debug, PartialEq)]
+pub struct Repository {
+	/// The url of the repository.
+	pub url: Url,
+	/// If applicable, the branch or tag to be used.
+	pub reference: Option<String>,
+	/// The name of a package within the repository. Defaults to the repository name.
+	pub package: String,
+}
+
+impl Repository {
+	/// Parses a url in the form of https://github.com/org/repository?package#tag into its component parts.
+	///
+	/// # Arguments
+	/// * `url` - The url to be parsed.
+	pub fn parse(url: &str) -> Result<Self, Error> {
+		let url = Url::parse(url)?;
+		let package = url.query();
+		let reference = url.fragment().map(|f| f.to_string());
+
+		let mut url = url.clone();
+		url.set_query(None);
+		url.set_fragment(None);
+
+		let package = match package {
+			Some(b) => b,
+			None => crate::GitHub::name(&url)?,
+		}
+		.to_string();
+
+		Ok(Self { url, reference, package })
+	}
+}
+
+/// Trait for observing status updates.
+pub trait Status {
```

Comment:
This should be in the crate root imo, its a general purpose trait for reporting status updates and not specific to a binary. Likewise the default impl for ().